File: grendel/sources/dog/util/Timer.java
/*
 * Timer.java
 *
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * The Original Code is Knife.
 *
 * The Initial Developer of the Original Code is dog.
 * Portions created by dog are
 * Copyright (C) 1998 dog <dog@dog.net.uk>. All
 * Rights Reserved.
 *
 * Contributor(s): n/a.
 *
 * You may retrieve the latest version of this package at the knife home page,
 * located at http://www.dog.net.uk/knife/
 */

package dog.util;

/**
 * A timer class that wakes up listeners after a specified number of milliseconds.
 *
 * @author dog@dog.net.uk
 * @version 0.3
 */
public final class Timer extends Thread {

	long time;
	long interval = 0;
	TimerListener listener;

	/**
	 * Constructs a timer.
	 */
	public Timer(TimerListener listener) {
		this(listener, 0, false);
	}
	
	/**
	 * Constructs a timer with the specified interval, and starts it.
	 */
	public Timer(TimerListener listener, long interval) {
		this(listener, interval, true);
	}
	
	/**
	 * Constructs a timer with the specified interval, indicating whether or not to start it.
	 */
	public Timer(TimerListener listener, long interval, boolean start) {
		this.listener = listener;
		this.interval = interval;
		time = System.currentTimeMillis();
		setDaemon(true);
		setPriority(Thread.MIN_PRIORITY);
		if (start)
			start();
	}

	// -- Accessor methods --

	/**
	 * Returns this timer's interval.
	 */
	public long getInterval() {
		return interval;
	}

	/**
	 * Sets this timer's interval.
	 */
	public void setInterval(long interval) {
		this.interval = interval;
	}

	/**
	 * Runs this timer.
	 */
	public void run() {
		boolean interrupt = false;
		while (!interrupt) {
			synchronized (this) {
				try {
					wait(interval);
				} catch (InterruptedException e) {
					interrupt = true;
				}
				listener.timerFired(new TimerEvent(this, interval));
			}
		}
	}

}

File: ef/Compiler/RegisterAllocator/Timer.cpp
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
 *
 * The contents of this file are subject to the Netscape Public License
 * Version 1.0 (the "NPL"); you may not use this file except in
 * compliance with the NPL.  You may obtain a copy of the NPL at
 * http://www.mozilla.org/NPL/
 *
 * Software distributed under the NPL is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the NPL
 * for the specific language governing rights and limitations under the
 * NPL.
 *
 * The Initial Developer of this code under the NPL is Netscape
 * Communications Corporation.  Portions created by Netscape are
 * Copyright (C) 1998 Netscape Communications Corporation.  All Rights
 * Reserved.
 */

#include "Fundamentals.h"
#include "HashTable.h"
#include "Timer.h"
#include "Pool.h"

static Pool pool;											// Pool for the Timer class.
static HashTable<TimerEntry*> timerEntries(pool);			// Timers hashtable.

const nTimersInABlock = 128;								// Number of timers in a block.
static PRTime *timers = new(pool) PRTime[nTimersInABlock];	// A block of timers.
static Uint8 nextTimer = 0;									// nextAvailableTimer.

//
// Calibrate the call to PR_Now().
//
static PRTime calibrate()
{
	PRTime t = PR_Now();
	PRTime& a = *new(pool) PRTime();

	// Call 10 times the PR_Now() function.
	a = PR_Now(); a = PR_Now(); a = PR_Now(); a = PR_Now(); a = PR_Now(); a = PR_Now(); 
	a = PR_Now(); a = PR_Now(); a = PR_Now(); a = PR_Now(); a = PR_Now(); a = PR_Now(); 
	t = (PR_Now() - t + 9) / 10;

	return t;
}

static PRTime adjust = calibrate();

//
// Return the named timer..
//
TimerEntry& Timer::getTimerEntry(const char* name)
{
	if (!timerEntries.exists(name)) {
		TimerEntry* newEntry = new(pool) TimerEntry();
		newEntry->accumulator = 0;
		newEntry->running = false;
		timerEntries.add(name, newEntry);
	}

	return *timerEntries[name];
}

//
// Return a reference to a new timer.
//
PRTime& Timer::getNewTimer()
{
	if (nextTimer >= nTimersInABlock) {
		timers = new(pool) PRTime[nTimersInABlock];
		nextTimer = 0;
	}
	return timers[nextTimer++];
}

static Uint32 timersAreFrozen = 0;

//
// Start the named timer.
//
void Timer::start(const char* name)
{
	if (timersAreFrozen)
		return;

	freezeTimers();

	TimerEntry& timer = getTimerEntry(name);
	PR_ASSERT(!timer.running);

	timer.accumulator = 0;
	timer.running = true;
	timer.done = false;

	unfreezeTimers();
}

//
// Stop the named timer.
//
void Timer::stop(const char* name)
{
	if (timersAreFrozen)
		return;

	freezeTimers();

	TimerEntry& timer = getTimerEntry(name);
	PR_ASSERT(timer.running);
	timer.running = false;
	timer.done = true;

	unfreezeTimers();
}

//
// Freeze all the running timers.
//
void Timer::freezeTimers()
{
	PRTime when = PR_Now() - adjust;

	if (timersAreFrozen == 0) {
		Vector<TimerEntry*> entries = timerEntries;
		Uint32 count = entries.size();

		for (Uint32 i = 0; i < count; i++) {
			TimerEntry& entry = *entries[i];
			if (entry.running) {
				entry.accumulator += (when - *entry.startTime);
			}
		}
	}
	timersAreFrozen++;
}
 
//
// Unfreeze all the running timers.
//
void Timer::unfreezeTimers()
{
	PR_ASSERT(timersAreFrozen != 0);
	timersAreFrozen--;

	if (timersAreFrozen == 0) {
		Vector<TimerEntry *> entries = timerEntries;
		Uint32 count = entries.size();
		
		PRTime& newStart = getNewTimer();
		
		for (Uint32 i = 0; i < count; i++) {
			TimerEntry& entry = *entries[i];
			if (entry.running) {
				entry.startTime = &newStart;
			}
		}
		
		newStart = PR_Now();
	}
}

//
// Print the named timer in the file f.
//
void Timer::print(FILE* f, const char *name)
{
	if (timersAreFrozen)
		return;

	freezeTimers();

	TimerEntry& timer = getTimerEntry(name);
  
	PR_ASSERT(timer.done);
	PRTime elapsed = timer.accumulator;

	if (elapsed >> 32) {
		fprintf(f, "[timer %s out of range]\n", name);
	} else {
		fprintf(f, "[%dus in %s]\n", Uint32(elapsed), name);
	}
	fflush(f);

	unfreezeTimers();
}

